proposals/NNNN-typed-throws-concurrency-library.md

```diff
@@ -0,0 +1,319 @@
+}
+```
+
+These two can be replaced with a single property:
+
+```swift
+var value: Success { get async throws(Failure) }
+```
+
+### Continuations
+
+Currently, there are two variants of each continuation creation method to
+accommodate for the throwing and non-throwing variant. Since, the natural
+spelling would be without the `Throwing` in the method name. We propose
+to add the following two new methods:
+
+```swift
+public func withCheckedContinuation<T, Failure: Error>(
+    function: String = #function,
+    _ body: (CheckedContinuation<T, Failure>) -> Void
+) async throws(Failure) -> T
+
+public func withUnsafeContinuation<T, Failure: Error>(
+  _ fn: (UnsafeContinuation<T, Failure>) -> Void
+) async throws(Failure) -> T
+```
+
+#### Task cancellation
+
+A few of the `Task` APIs are documented to only throw `CancellationError` and
+can adopt typed throws. For example, `checkCancellation`:
+
+```swift
+public static func checkCancellation() throws(CancellationError)
+```
+
+Similarly, the `sleep` APIs will only throw on cancellation:
+
+```swift
+public static func sleep<C: Clock>(
+  until deadline: C.Instant,
+  tolerance: C.Instant.Duration? = nil,
+  clock: C = ContinuousClock()
+) async throws(CancellationError)
+
+public static func sleep<C: Clock>(
+  for duration: C.Instant.Duration,
+  tolerance: C.Instant.Duration? = nil,
+  clock: C = ContinuousClock()
+) async throws(CancellationError)
+```
+
+Additionally, the `ContinuousClock` and `SuspendingClock` `sleep` methods will also
+adopt typed throws:
+
+```swift
+public func sleep(
+  until deadline: Instant, tolerance: Swift.Duration? = nil
+) async throws(CancellationError)
+```
+
+Lastly, the `withTaskCancellationHandler` needs a new variant
+that supports a typed error `operation` closure:
+
+```swift
+public func withTaskCancellationHandler<T, Failure: Error>(
+  operation: () async throws(Failure) -> T,
+  onCancel handler: @Sendable () -> Void
+) async rethrows -> T
+```
+
+### Task groups
+
+The task group APIs also need to gain support for typed throws. The important
+detail here is that there are two layers of failures involved. First, the
+failures produced by the child task themselves. Secondly, the failure returned
+by `withThrowingTaskGroup` API itself. We propose to add the following API:
+
+```swift
+public func withTaskGroup<ChildTaskResult, ChildTaskFailure: Error, GroupResult, GroupFailure: Error>(
+  of childTaskResultType: ChildTaskResult.Type,
+  childTaskFailureType: ChildTaskFailure.Type,
+  returning returnType: GroupResult.Type = GroupResult.self,
+  body: (inout ThrowingTaskGroup<ChildTaskResult, ChildTaskFailure>) async throws(GroupFailure) -> GroupResult
+) async rethrows -> GroupResult
+```
+
+Additionally, we propose to add typed throws to following the
+`ThrowingTaskGroup` APIs.
+
+```swift
+public mutating func next() async throws(Failure) -> ChildTaskResult?
+public mutating func waitForAll() async throws(Failure)
```

Comment:
This is confusing to read since above we don't talk about Failure, but I know that this means:

```suggestion
public struct TaskGroup<ChildTaskSuccess, ChildTaskFailure> {
  public mutating func next() async throws(ChildTaskFailure)
  public mutating func waitForAll() async throws(ChildTaskFailure)
}

```

Reply:
And don't forget the `TaskGroup.Iterator`:

```
struct ThrowingTaskGroup {
  public struct Iterator: AsyncIteratorProtocol {
    public mutating func next() async throws(ChildTaskFailure) -> Element? 
  }
}
```

let's add that here as well